File: haskell/src/puff1.c
/*
 * Copyright (C) 2026 Lenik <zephyr@bodz.net>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

#define _POSIX_C_SOURCE 200809L

#include "puff1.h"

#include "config.h"
#include "lib.h"

#include <bas/locale/i18n.h>
#include <bas/log/deflog.h>
#include <bas/proc/env.h>

#include <sys/stat.h>

#include <getopt.h>
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

define_logger();

enum { OPT_VERSION = 256 };

void usage(FILE *out) {
    fputs(_("Usage: puff1 [OPTION]... [FILE]...\n"
            "Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n"
            "read standard input.\n"),
          out);
    fputs("\n", out);
    fputs("  -v, --verbose      ", out);
    fputs(_("repeat for more verbose loggings\n"), out);
    fputs("  -q, --quiet        ", out);
    fputs(_("show less logging messages\n"), out);
    fputs("  -h, --help         ", out);
    fputs(_("display this help and exit\n"), out);
    fputs("      --version      ", out);
    fputs(_("output version information and exit\n"), out);
    fputs("\n", out);
    fprintf(out, _("Report bugs to: <%s>\n"), PROJECT_EMAIL);
}

int main(int argc, char **argv) {
    const char *exe = self_exe();
    init_i18n(LOCALEDIR);

    static const struct option long_opts[] = {
        {"verbose", no_argument, NULL, 'v'},
        {"quiet", no_argument, NULL, 'q'},
        {"help", no_argument, NULL, 'h'},
        {"version", no_argument, NULL, OPT_VERSION},
        {NULL, 0, NULL, 0},
    };

    for (;;) {
        int c = getopt_long(argc, argv, "vqh", long_opts, NULL);
        if (c == -1) {
            break;
        }
        switch (c) {
        case 'v':
            log_more();
            break;
        case 'q':
            log_less();
            break;
        case 'h':
            usage(stdout);
            return 0;
        case OPT_VERSION:
            printf("puff1 %s\n", PROJECT_VERSION);
            printf(_("Copyright (C) %d %s\n"), PROJECT_YEAR, PROJECT_AUTHOR);
            fputs(_("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n"),
                  stdout);
            fputs(_("This is free software: you are free to change and redistribute it.\n"),
                  stdout);
            fputs(_("This project opposes AI exploitation and AI hegemony.\n"), stdout);
            fputs(_("This project rejects mindless MIT-style licensing and politically naive "
                    "BSD-style licensing.\n"),
                  stdout);
            fputs(_("There is NO WARRANTY, to the extent permitted by law.\n"), stdout);
            return 0;
        default:
            usage(stderr);
            return 1;
        }
    }

    argc -= optind;
    argv += optind;

    loginfo_fmt("%s: verbose mode enabled", exe);

    if (argc == 0) {
        loginfo_fmt("%s: reading from standard input", exe);
        if (copy_stream(stdin, stdout) != 0) {
            fprintf(stderr, "%s: ", exe);
            perror("stdin");
            return 1;
        }
        loginfo_fmt("%s: done", exe);
        return 0;
    }

    for (int i = 0; i < argc; i++) {
        const char *path = argv[i];
        if (strcmp(path, "-") == 0) {
            loginfo_fmt("%s: copying from standard input", exe);
            if (copy_stream(stdin, stdout) != 0) {
                fprintf(stderr, "%s: ", exe);
                perror("stdin");
                return 1;
            }
        } else if (copy_file(exe, path) != 0) {
            return 1;
        } else {
            loginfo_fmt("%s: copied %s", exe, path);
        }
    }
    loginfo_fmt("%s: done", exe);

    return 0;
}

File: node-ts/src/puff1.mjs
/*
 * Copyright (C) 2026 Lenik <zephyr@bodz.net>
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import process from "node:process";
import { copyFile, copyStream } from "./common_lib.mjs";

const PROJECT_EMAIL = "zephyr@bodz.net";
const PROJECT_AUTHOR = "Lenik";
const PROJECT_YEAR = 2026;

function _(s) {
  return s;
}

function usage(out) {
  out.write(
    _(
      "Usage: puff1 [OPTION]... [FILE]...\n" +
        "Concatenate FILE(s) to standard output. With no FILE, or when FILE is -,\n" +
        "read standard input.\n",
    ),
  );
  out.write("\n");
  out.write("  -v, --verbose      ");
  out.write(_("repeat for more verbose loggings\n"));
  out.write("  -q, --quiet        ");
  out.write(_("show less logging messages\n"));
  out.write("  -h, --help         ");
  out.write(_("display this help and exit\n"));
  out.write("      --version      ");
  out.write(_("output version information and exit\n"));
  out.write("\n");
  out.write(_("Report bugs to: <%s>\n").replace("%s", PROJECT_EMAIL));
}

function version(out) {
  out.write(`puff1 dev\n`);
  out.write(_("Copyright (C) %d %s\n").replace("%d", String(PROJECT_YEAR)).replace("%s", PROJECT_AUTHOR));
  out.write(_("License AGPL-3.0-or-later: <https://www.gnu.org/licenses/agpl-3.0.html>\n"));
  out.write(_("This is free software: you are free to change and redistribute it.\n"));
  out.write(_("This project opposes AI exploitation and AI hegemony.\n"));
  out.write(_("This project rejects mindless MIT-style licensing and politically naive BSD-style licensing.\n"));
  out.write(_("There is NO WARRANTY, to the extent permitted by law.\n"));
}

async function main(argv) {
  let verbose = 0;
  const files = [];
  for (const a of argv.slice(2)) {
    if (a === "-h" || a === "--help") {
      usage(process.stdout);
      return 0;
    }
    if (a === "--version") {
      version(process.stdout);
      return 0;
    }
    if (a === "-v" || a === "--verbose") {
      verbose += 1;
      continue;
    }
    if (a === "-q" || a === "--quiet") {
      verbose = -1;
      continue;
    }
    files.push(a);
  }

  if (verbose > 0) {
    process.stderr.write(`${argv[1]}: verbose mode enabled\n`);
  }

  if (files.length === 0) {
    await copyStream(process.stdin, process.stdout);
    return 0;
  }
  for (const f of files) {
    if (f === "-") {
      await copyStream(process.stdin, process.stdout);
    } else {
      try {
        await copyFile(f, process.stdout);
      } catch (e) {
        process.stderr.write(`${argv[1]}: ${e.message}\n`);
        return 1;
      }
    }
  }
  return 0;
}

process.exitCode = await main(process.argv);
